### ISSUE TITLE
Using ESBuild

### ISSUE THREAD

**Issue by NullVoxPopuli** [microsoft/monaco-editor#2411]
![image](https://user-images.githubusercontent.com/199018/112568713-4a4da200-8db9-11eb-9e9d-ca871314a6ca.png)

👋 I know there are a number of issues asking "how to integrate with X" or "how to use with technology Y", so I wanted to share something that I got working without any fuss. The existing examples were helpful in some of this, but I hope this is helpful for intrepid adventurers!

I should note that at the time of writing ESBuild considers itself experimental, and this may change after posting.

The process is mainly two parts:
 - bundling the workers
 - loading the editor / telling it where the workers are
 
 ## Bundling the workers:
 
 Since I'm using [EmberJS](https://emberjs.com), there are some extra bits to integrate ESBuild into the existing build tools (mainly broccoli-funnel, and broccoli-merge-trees) -- these can be ignored for non-ember projects.

Source: https://github.com/NullVoxPopuli/limber/blob/main/lib/monaco.js

```js
// <git root>/lib/monaco.js
'use strict';

const path = require('path');
const os = require('os');
const fs = require('fs');
const esbuild = require('esbuild');

const Funnel = require('broccoli-funnel');
const mergeTrees = require('broccoli-merge-trees');

// const monacoLocation = path.join('node_modules', 'monaco-editor', 'esm', 'vs');
const cssLocation = path.join('node_modules', 'monaco-editor', 'min', 'vs', 'editor');

const workers = {
  base: 'node_modules/monaco-editor/esm/vs/editor/editor.main.js',
  editor: 'node_modules/monaco-editor/esm/vs/editor/editor.worker.js',
  json: 'node_modules/monaco-editor/esm/vs/language/json/json.worker.js',
  css: 'node_modules/monaco-editor/esm/vs/language/css/css.worker.js',
  html: 'node_modules/monaco-editor/esm/vs/language/html/html.worker.js',
  ts: 'node_modules/monaco-editor/esm/vs/language/typescript/ts.worker.js',
};

function buildWorkers({ isProduction }) {
  let buildDir = fs.mkdtempSync(path.join(os.tmpdir(), 'monaco--workers'));

  esbuild.buildSync({
    loader: { '.ts': 'ts', '.js': 'js' },
    entryPoints: [
      workers.editor,
      workers.json,
      workers.css,
      workers.html,
      workers.ts,
      workers.base,
    ],
    bundle: true,
    outdir: buildDir,
    format: 'iife',
    minify: isProduction,
    sourcemap: false,
  });

  return new Funnel(buildDir, {
    destDir: 'monaco/',
    exclude: ['*.nls.*'],
  });
}

function monacoFunnel(options) {
  let src = buildWorkers(options);

  // let src = new Funnel(monacoLocation, {
  //   destDir: 'monaco/',
  //   // exclude: ['*.nls.*'],
  // });

  let css = new Funnel(cssLocation, {
    destDir: 'monaco/',
    include: ['*.css'],
  });

  return mergeTrees([src, css]);
}

module.exports = {
  monacoFunnel,
};
```

And then in my `ember-cli-build.js` file:

```js
const { monacoFunnel } = require('./lib/monaco');
const { Webpack } = require('@embroider/webpack');

module.exports = function (defaults) {
  let environment = EmberApp.env();
  let isProduction = environment === 'production';

  let app = new EmberApp(defaults, { /* custom config */ });


  return require('@embroider/compat').compatBuild(app, Webpack, {
    extraPublicTrees: [ monacoFunnel({ isProduction }) ],
  });
};
```

## Using the Editor

Source: https://github.com/NullVoxPopuli/limber/blob/main/app/modifiers/monaco.js#L29

```ts
import { modifier } from 'ember-could-get-used-to-this';
import * as monaco from 'monaco-editor';

const CSS = '/monaco/editor.main.css';
const OPTIONS = {
   /* static options */
};

export default modifier((element, [value, updateText]) => {
  configureWorkerPaths();
  insertStyles();

  let editor = monaco.editor.create(element, {
    ...OPTIONS,
    value,
  });

  editor.onDidChangeModelContent(() => {
    updateText(editor.getValue());
  });

  return () => editor.dispose();
});

async function insertStyles() {
  let hasStyles = document.querySelector(`link[href="${CSS}]"`);

  if (hasStyles) return;

  let element = document.createElement('link');

  element.setAttribute('rel', 'stylesheet');
  element.setAttribute('href', CSS);

  document.body.appendChild(element);
  await Promise.resolve();
}
function configureWorkerPaths() {
  if (window.MonacoEnvironment) return;

  window.MonacoEnvironment = {
    getWorkerUrl: function (moduleId, label) {
      if (label === 'json') {
        return '/monaco/language/json/json.worker.js';
      }

      if (label === 'css' || label === 'scss' || label === 'less') {
        return '/monaco/language/css/css.worker.js';
      }

      if (label === 'html' || label === 'handlebars' || label === 'razor') {
        return '/monaco/language/html/html.worker.js';
      }

      if (label === 'typescript' || label === 'javascript') {
        return '/monaco/language/typescript/ts.worker.js';
      }

      return '/monaco/editor/editor.worker.js';
    },
  };
}
```

And then, in Ember at least, the final user-land API is:

```hbs
<div {{monaco this.value this.updateValue}}></div>
```

**Comment by alexdima** [comment]
Thank you for your example! I've added an esbuild sample at https://github.com/microsoft/monaco-editor/tree/main/samples/browser-esm-esbuild

**Comment by Prinzhorn** [comment]
@alexdima unfortunately the example appears broken:

1. No `dependencies`/`devDependencies` in [package.json](https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/samples/browser-esm-esbuild/package.json) (monaco-editor, esbuild). But I assume the samples are not self-contained and require `npm i` at the root.  If I do that, `npm run build` fails with "error: Could not resolve "../node_modules/monaco-editor/esm/vs/editor/editor.worker.js"" because I also need to install `monaco-editor` in the samples folder. It would be great if the samples could be self contained to that they can be easily ejected from the main repo. I guess `../../build/utils` could then be replaced with `rimraf` or whatever fits.
2. More importantly: `build.js` [deletes `dist`](https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/samples/browser-esm-esbuild/build.js#L5) including `dist/index.html`, which is the entry file and different from the root `index.html`. But it actually doesn't, because `removeDir` [already does `path.join(REPO_ROOT, _dirPath)`](https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/build/utils.js#L66), so i end up with a non-existing path (joining one absolute path with another, basically `/home/alex/**/home/alex/**`).

TLDR:

1. Need to `npm i` in three different folders
2. `build.js` and the entire `index.html` business doesn't appear to do anything useful

**Comment by Prinzhorn** [comment]
@alexdima the sample appears broken, this doesn't actually delete anything: https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/samples/browser-esm-esbuild/build.js#L5 because `removeDir` will join the absolute path again https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/build/utils.js#L66, resulting in a wrong path (joining two absolute paths). E.g. for me this ends up being `/tmp/monaco-editor/tmp/monaco-editor/samples/browser-esm-esbuild/dist` having cloned into `/tmp/monaco-editor`.

If it _would_ correctly delete `dist`, then it would also remove `index.html` and you couldn't actually run the sample. So I guess it accidentally "works" right now. Other samples have a `src` folder, which also contains the `index.html`. The `build.js` could copy it over.

**Comment by hediet** [comment]
@Prinzhorn it would be awesome if you could open a new issue for that!
